### ISSUE TITLE
Failed testcase is marked as SUCCESS after using RetryAnalyzer 

### ISSUE THREAD

**Issue by hemal7735** [cbeust/testng#2272]
**Apologies if this issue creates a noise!**

### TestNG Version - 6.9.6
> Note: only the latest version is supported

### Expected behavior
I'm using `IRetryAnalyzer` to retry the simple test which asserts `(1 == 2)`. The test case is The first 2 runs are marked `SKIP` as per https://github.com/cbeust/testng/pull/698.
**However, the third(and final) run is marked as SUCCESS.**

### Actual behavior
**However, the third(and final) run should be marked as FAILURE as (1==2) is a failed condition.**

### Is the issue reproducible on a runner?

- [ ] Shell
- [X] Maven
- [ ] Gradle
- [ ] Ant
- [ ] Eclipse
- [X] IntelliJ
- [ ] NetBeans

### Test case sample
> Please, share the test case (as small as possible) which shows the issue

**RetryAnalyzer.java**
```java
public class RetryAnalyzer implements IRetryAnalyzer {

    private int count = 0;
    private static int maxTry = 3;

    @Override
    public boolean retry(ITestResult result) {
        if (!result.isSuccess()) {                      
            if (count < maxTry) {                       
                count++;                                
                result.setStatus(ITestResult.FAILURE);  
                return true;                            
            } else {
                result.setStatus(ITestResult.FAILURE);  
            }
        } else {
            result.setStatus(ITestResult.SUCCESS);
        }
        return false;
    }
}
```
**FooTest.java**
```java
public class FooTest {
    @Test(retryAnalyzer = RetryAnalyzer.class)
    public void testRetry() {
         assertEquals(1,2,"Addition Problem! 1 can't be 2!\n");
    }
}
```

### My Analysis:
Although I'm not a TestNG expert, I have done some analysis. Please find the summary below.
All the code lies in `Invoker.java`.

1. Before #698 we were considering FAILURE as a FAILURE. Post this PR, we also started resetting retryable FAILURE as SKIP. However, we don't increase failure count, which makes sense as we haven't marked the test-case failure, but skipped it.
```java
IRetryAnalyzer retryAnalyzer = testMethod.getRetryAnalyzer();
      boolean willRetry = retryAnalyzer != null && status == ITestResult.FAILURE && failure.instances != null && retryAnalyzer.retry(testResult);

      if (willRetry) {
        resultsToRetry.add(testResult);
        failure.instances.add(testResult.getInstance());
        testResult.setStatus(ITestResult.SKIP);
      } else {
        testResult.setStatus(status);
        if (status == ITestResult.FAILURE && !handled) {
          handleException(ite, testMethod, testResult, failure.count++);
        }
      }
      if (collectResults) {
        // Collect the results
        collectResults(testMethod, Collections.singleton(testResult));
//        if (triggerListeners && status != ITestResult.SUCCESS) {
//          runTestListeners(testResult);
//        }
      }
    } // for results

    removeResultsToRetryFromResult(resultsToRetry, result, failure);
```
2. Now inside `removeResultsToRetryFromResult` we decreasing the `failure.count` even for the tests which are marked `SKIP` with #698 change. This makes `failure.count` negative, which is incorrect.
```java
private void removeResultsToRetryFromResult(List<ITestResult> resultsToRetry,
                                              List<ITestResult> result, FailureContext failure) {
    if (resultsToRetry != null) {
      for (ITestResult res : resultsToRetry) {
        result.remove(res);
        failure.count--;
      }
    }
  }
```
3. Next, after the final retry we won't mark the testcase as `SKIP`, and will call `handleException`. 
Since, `failure.count` is negative and `successPercentage` is 100, the below conditions passes the test.
```java
if(failureCount < numberOfTestsThatCanFail) {
      testResult.setStatus(ITestResult.SUCCESS_PERCENTAGE_FAILURE);
    }
```

```java
private void handleException(Throwable throwable,
                               ITestNGMethod testMethod,
                               ITestResult testResult,
                               int failureCount) {
    if (throwable != null) {
      testResult.setThrowable(throwable);
    }
    int successPercentage= testMethod.getSuccessPercentage();
    int invocationCount= testMethod.getInvocationCount();
    float numberOfTestsThatCanFail= ((100 - successPercentage) * invocationCount) / 100f;

    if(failureCount < numberOfTestsThatCanFail) {
      testResult.setStatus(ITestResult.SUCCESS_PERCENTAGE_FAILURE);
    }
    else {
      testResult.setStatus(ITestResult.FAILURE);
    }
  }
```

**Comment by krmahadevan** [comment]
@hemal7735 - Please try using the latest released version of TestNG `7.1.0` and post back your experience/findings. You are using a very old version of TestNG.

**Comment by hemal7735** [comment]
Hi @krmahadevan , looks like it is fixed in `7.1.0`.
However, now I'm interested in knowing about how the team identified this and fixed it 🤔 
Could you share any docs if you have?

**Comment by krmahadevan** [comment]
@hemal7735 - No I don't have any docs. Obviously there must have been a bug that someone raised and it got fixed.

I would suggest that you work backwards on the changes (https://github.com/cbeust/testng/blob/master/CHANGES.txt) find issues that are related to retry analyser and then check which of them tracks your problem and look at the corresponding PR that fixed it.

Closing this issue.
